daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala

```diff
@@ -128,24 +120,20 @@ object DataProcessor {
  * back-end runtime.
  */
 class DataProcessor private (
-  //
-  // Once the deprecated API goes away, these var can become val.
-  //
-  var ssrd: SchemaSetRuntimeData,
-  var tunables: DaffodilTunables, // Compiler-set tunables
-  var variableMap: VariableMap,
+  val ssrd: SchemaSetRuntimeData,
+  val tunables: DaffodilTunables, // Compiler-set tunables
+  val variableMap: VariableMap,
   //
   // The below do not need to be transient
   // because this object itself isn't serialized. A SerializableDataProcessor is.
   // The values these will have (since this is a base class) are the correct default values that we want
   // back when the object is re-initialized.
   //
-  protected var areDebugging : Boolean,
-  protected var optDebugger : Option[Debugger],
+  protected val areDebugging : Boolean,
+  protected val optDebugger : Option[Debugger],
   var validationMode: ValidationMode.Type,
   private var externalVars: Queue[Binding])
```

Comment:
Can `validationMode` and `externalVars` become  a val too like above? We don't have the set* functions anymore so nothing should mutate this I think?

Reply:
Right above this we do `val dpToSave = dp.copy()`. We could change that to something like his:
```scala
val dpToSave = this.copy(
  externalVars = Queue.empty[Binding],
  validationMode = ValidationMode.Off
)
```
That's a fairly small change and I think makes our DataProcessor fully immutable.

Reply:
That makes sense I will throw that in

Reply:
@stevedlawrence That look good? So since I have two approvals, I should wait till the CI passed, then squash my commits, push up the squashed commit and then merge once CI passes?

Reply:
Yes, all the latest changes look good (thanks, Shane and Steve, for adding them).  Yes, CI has passed so you can squash your commits, push a single squashed commit, and merge (if you have write permission).

Reply:
Yes, the commit message can reuse your nicely written PR description (updated for the latest changes).  That would be very nice for git history.